src/Functions/tokens.cpp

```diff
@@ -0,0 +1,166 @@
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/IFunction.h>
+#include <Interpreters/Context_fwd.h>
+#include <Interpreters/ITokenExtractor.h>
+
+#include "config.h"
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+}
+
+class FunctionTokens : public IFunction
+{
+    static constexpr size_t arg_value = 0;
+    static constexpr size_t arg_tokenizer = 1;
+    static constexpr size_t arg_ngrams = 2;
+    static constexpr size_t arg_chinese_granularity = arg_ngrams;
+
+public:
+    static constexpr auto name = "tokens";
+
+    static FunctionPtr create(ContextPtr) { return std::make_shared<FunctionTokens>(); }
+
+    String getName() const override { return name; }
+    size_t getNumberOfArguments() const override { return 0; }
+    bool isVariadic() const override { return true; }
+    bool useDefaultImplementationForConstants() const override { return true; }
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        FunctionArgumentDescriptors mandatory_args{
+            {"value", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isStringOrFixedString), nullptr, "String or FixedString"}};
+
+        FunctionArgumentDescriptors optional_args;
+
+        if (arguments.size() > 1)
+        {
+            optional_args.emplace_back("tokenizer", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isString), isColumnConst, "String");
+
+            validateFunctionArguments(*this, {arguments[arg_value], arguments[arg_tokenizer]}, mandatory_args, optional_args);
+
+            if (arguments.size() == 3)
+            {
+                const auto tokenizer = arguments[arg_tokenizer].column->getDataAt(0).toString();
+
+                if (tokenizer == NgramTokenExtractor::getExternalName())
+                    optional_args.emplace_back("ngrams", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isUInt8), isColumnConst, "UInt8");
+#if USE_CPPJIEBA
+                else if (tokenizer == ChineseTokenExtractor::getExternalName())
+                    optional_args.emplace_back("chinese_granularity", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isString), isColumnConst, "String");
+#endif
+            }
+        }
+
+        validateFunctionArguments(*this, arguments, mandatory_args, optional_args);
+
+        return std::make_shared<DataTypeArray>(std::make_shared<DataTypeString>());
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    {
+        auto col_input = arguments[arg_value].column;
+        auto col_result = ColumnString::create();
+        auto col_offsets = ColumnArray::ColumnOffsets::create();
+
+        if (input_rows_count == 0)
```

Comment:
That shortcut is important. Otherwise below code may read from empty columns and sanitizer builds will complain.